zetaclient/zetacore/background_threads.go

```diff
@@ -0,0 +1,162 @@
+package zetacore
+
+import (
+	"context"
+	"time"
+
+	"cosmossdk.io/errors"
+
+	"github.com/zeta-chain/zetacore/pkg/retry"
+	observertypes "github.com/zeta-chain/zetacore/x/observer/types"
+	zctx "github.com/zeta-chain/zetacore/zetaclient/context"
+)
+
+// HandleTSSUpdate is a background thread that listens for TSS updates; it returns when the TSS address is updated
+func (c *Client) HandleTSSUpdate(ctx context.Context) error {
```

Comment:
- Let's create a `zetaclient/maintenance` package and move TSS chores there. `zetaclient/zetacore/client.go` should be a client (i.e. req/res only), not a worker (there's 1 worker there atm which we need to move to orchestrator).
- Naming:
   - `(*TSSChore).WaitForUpdate(ctx) error` - blocking call. Returns `nil` on change
   - `(*TSSChore).WaitForNewTSSGeneration(ctx) error` - blocking call. Returns `nil` on change
   - `(*TSSChore).WaitForNewKeygen(ctx) error` - blocking call. Returns `nil` on change

Reply:
Also, the logic with cancel function is confusing, it's hard to grasp who actually calls cancel. I would not use `bg.WithCancel()` right now as well as this: (start.go)
```go
	// Create a notification context for background threads.These threads are responsible for sending shutdown signals to the main thread.
	notifyCtx, cancel := context.WithCancelCause(ctx)

	// Start background threads
	defer startBackgroundThreads(notifyCtx, cancel, zetacoreClient, masterLogger)
```

How I envision it:
```go
tssChores := maintenance.NewTSSChores(func(){
   // callback that is being invokes if we detect any changes in `Wait*` calls
   sigChannel <-  syscall.SIGINT
})


// invokes bg task
tssChores.Start(ctx)

```

and then we can simply implement bg opt `bg.OnComplete(callback)`








**Note**: Orchestrator is responsible for (de)provisioning observers/signers, no need to manually stop observers in start.go as it should happen within orchestrator's lifecycle.

Reply:
I agree that using bg.Cancel() is not the cleanest way to do this, but the arch mentioned in this comment should be part of a separate refactor .
In my mind , background tasks is fine , as a package , We should ideally have two types of tasks , ones that read only and ones that write something back to the parent thread (naming can be decided later ), that way it can be more general , rather that adding something like `TSSChores` which is too specific .

This PR tries to keep changes to the zetaclient minimal .

Reply:
Summarizing points discussed and additional thoughts . From what I understand, there are two main things that you are suggesting 
- Use a callback instead of the cancel function: I agree that cancel function is not the ideal option, but it provides some advantages, such as canceling all sub-routines when the parent is canceled. The cancelFunction provided here is actually the callback (renamed so that the terms make more sense ) 

- Made the functions as part of a separate struct and not `client` :  I don't see a lot of advantage of doing this , considering we are the functions that these routines use are all part of the client right now . Unless we are planning to modify functions such as `GetTss(ctx)` to `GetTss(ctx,client)`

Reply:
Several points from my side:
- By implementing reusable code components, we can simplify the mental model behind the project.
- In order to keep the codebase maintainable, we need to reduce the technical debt or at least stop increasing it. In my point of view, adding net-new worker logic in the client's package (a package that should handle req/res only) is indeed a technical debt increase due to making a "shortcut."
- `bg` package can and should be improved in the future PRs, but for this PR,`bg.OnSuccess(func())` is more than enough (IMO)

Reply:
I implemented proposed changes, feel free to ask any questions:
- https://github.com/zeta-chain/node/pull/2538/commits/f993da5aa5db578556741711bbab7f9ae8b732b7
- https://github.com/zeta-chain/node/pull/2538/commits/7c113a5059da6ce9e6e7a87dc4c1fcecaeacbfe8

TLDR:
```go
// cmd/zetaclient/start.go
func start(_ *cobra.Command, _ []string) error {
  // ...
  
  // Creating a channel to listen for os signals (or other signals)
  signalChannel := make(chan os.Signal, 1)
  signal.Notify(signalChannel, syscall.SIGINT, syscall.SIGTERM)
  
  // Maintenance workers ============
  maintenance.NewTSSListener(zetacoreClient, masterLogger).Listen(ctx, func() {
      masterLogger.Info().Msg("TSS listener received an action to shutdown zetaclientd.")
      signalChannel <- syscall.SIGTERM
  })
  
  sig := <-signalChannel
  startLogger.Info().Msgf("Stop signal received: %q", sig)
  
  zetacoreClient.Stop()
}
```

cc @lumtis